### PR TITLE
Wallabag remove from history setting

### PR DESCRIPTION
**Pull request by mwoz123**
Solves: #5045

Default is (as previous) off

Zrzut ekranu z 2020-04-14 13-05-40

---
This change is [Reviewable](https://reviewable.io/reviews/koreader/koreader/6048)

### CODE REVIEW

File: plugins/wallabag.koplugin/main.lua [hunk 1]
```diff
@@ -1019,4 +1033,14 @@ function Wallabag:onSynchronizeWallabag()
     return true
 end
 
+
+function Wallabag:onCloseDocument()
+    if self.remove_from_korader_history then
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua:1038]: Surely only if status == "complete" or status == "abandoned"?

File: plugins/wallabag.koplugin/main.lua [hunk 2]
```diff
@@ -81,6 +82,7 @@ function Wallabag:init()
     if self.wb_settings.data.wallabag.articles_per_sync ~= nil then
         self.articles_per_sync = self.wb_settings.data.wallabag.articles_per_sync
     end
+    self.remove_from_korader_history = self.wb_settings.data.wallabag.remove_from_korader_history or false
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua:85]: Setting aside the typo, the name strikes me as rather verbose.

```suggestion
    self.remove_from_history = self.wb_settings.data.wallabag.remove_from_history or false
```

File: plugins/wallabag.koplugin/main.lua [hunk 3]
```diff
@@ -1019,4 +1033,14 @@ function Wallabag:onSynchronizeWallabag()
     return true
 end
 
+
+function Wallabag:onCloseDocument()
+    if self.remove_from_korader_history then
+        local document_full_path = self.ui.document.file
+        if  document_full_path and self.directory and self.directory == string.sub(document_full_path, 1, string.len(self.directory)) then
```
Comment (Frenzie) [inline plugins/wallabag.koplugin/main.lua:1040]:
```suggestion
        if document_full_path and self.directory and self.directory == string.sub(document_full_path, 1, string.len(self.directory)) then
```

File: plugins/wallabag.koplugin/main.lua (hunk 2 above)
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:85]: Isn't it confusing with wallabag history (if it exists)?
You sure you want shorter?

File: plugins/wallabag.koplugin/main.lua [hunk 4]
```diff
@@ -965,7 +978,8 @@ function Wallabag:saveSettings()
         is_archiving_deleted  = self.is_archiving_deleted,
         is_auto_delete        = self.is_auto_delete,
         is_sync_remote_delete = self.is_sync_remote_delete,
-        articles_per_sync     = self.articles_per_sync
+        articles_per_sync     = self.articles_per_sync,
+        remove_from_korader_history = self.remove_from_korader_history
```
[the remark this answers is not in the document]
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:982]: Sure

File: plugins/wallabag.koplugin/main.lua (hunk 1 above)
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:1038]: Ok.
I'll search where I can find 'status'

Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:1038]: ok got it:

status = docinfo.data.summary.status
 local docinfo = DocSettings:open(entry_path)
;)

File: plugins/wallabag.koplugin/main.lua [hunk 5]
```diff
@@ -238,6 +240,17 @@ function Wallabag:addToMainMenu(menu_items)
                             self:saveSettings()
                         end,
                     },
+                    {
+                        text = _("Remove from KOReader history"),
```
[the remark this answers is not in the document]
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:244]: As above:
Are you sure if people won't think about "wallabag history" (and will be confused by that) as all the rest settings relates to wallabag?

[the remark this answers is not in the document]
Comment (pazos) [reply plugins/wallabag.koplugin/main.lua:244]: there's no *KOReader* vs *wallabag* history, so "remove from history" is enough to understand the option.

[the remark this answers is not in the document]
Comment (Frenzie) [reply plugins/wallabag.koplugin/main.lua:244]: It's probably to say something like "Remove finished articles from history" but I don't think adding "KOReader" aids much if any in clarity.

[the remark this answers is not in the document]
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua:244]: done

File: plugins/wallabag.koplugin/main.lua (hunk 1 above)
Comment (mwoz123) [reply plugins/wallabag.koplugin/main.lua]: @Frenzie - can you help with this status . where I can find it?
It's not in `docinfo.data`  (`DocSettings:open(self.ui.document.file).data`)
for both 100% read and unfinished

Logs: 

```
04/14/20-19:59:44 DEBUG Wallabag: self.ui.document.file = /mnt/Dane/koreader/test/sample.pdf
04/14/20-19:59:44 DEBUG data is read from  /mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua
04/14/20-19:59:44 DEBUG Wallabag: DocSettings:open(self.ui.document.file)  = {
    ["filepath"] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
    ["data"] = {
        ["kopt_word_spacing"] = -0.2,
        ["kopt_contrast"] = 1,
        ["kopt_text_wrap"] = 0,
        ["render_mode"] = 0,
        ["highlight_drawer"] = "lighten",
        ["kopt_screen_mode"] = "portrait",
        ["gamma"] = 1,
        ["kopt_page_margin"] = 0.1,
        ["screen_mode"] = "portrait",
        ["kopt_font_fine_tune"] = 0.05,
        ["inverse_reading_order"] = false,
        ["show_overlap_enable"] = false,
        ["kopt_defect_size"] = 1,
        ["stats"] = {
            ["performance_in_pages"] = {},
            ["md5"] = "dabfd5be70e60d0d464b64ded495a3be",
            ["authors"] = "",
            ["total_time_in_sec"] = 0,
            ["notes"] = 0,
            ["title"] = "sample",
            ["highlights"] = 0,
            ["pages"] = 45,
            ["series"] = "",
            ["language"] = ""
        },
        ["doc_props"] = {
            ["creationDate"] = "D:20140516224813+08'00'",
            ["keywords"] = "",
            ["producer"] = "itext-paulo-155 (itextpdf.sf.net-lowagie.com)",
            ["authors"] = "",
            ["author"] = "",
            ["subject"] = "",
            ["title"] = "sample",
            ["description"] = "",
            ["modDate"] = "D:20140516224813+08'00'",
            ["creator"] = "pdftk 2.01 - www.pdftk.com",
            ["series"] = "",
            ["language"] = ""
        },
        ["kopt_auto_straighten"] = 0,
        ["config_panel_index"] = 1,
        ["highlights_imported"] = true,
        ["kopt_page_scroll"] = 1,
        ["kopt_detect_indent"] = 1,
        ["page_positions"] = {
            [2] = 0.69696969696969,
            [1] = 0.85353535353535
        },
        ["kopt_quality"] = 1,
        ["kopt_writing_direction"] = 0,
        ["flipping_zoom_mode"] = "page",
        ["kopt_page_gap_height"] = 8,
        ["highlight"] = {},
        ["bookmarks"] = {},
        ["readermenu_tab_index"] = 7,
        ["kopt_doc_language"] = "eng",
        ["zoom_mode"] = "pagewidth",
        ["partial_md5_checksum"] = "f40c4da7f2861191c5ec19de4d80894f",
        ["kopt_page_opt"] = 0,
        ["disable_fuzzy_search"] = false,
        ["kopt_font_size"] = 1,
        ["kopt_full_screen"] = 1,
        ["kopt_trim_page"] = 1,
        ["last_page"] = 2,
        ["rotation_mode"] = 0,
        ["kopt_max_columns"] = 2,
        ["kopt_justification"] = 3,
        ["highlight_disabled"] = false,
        ["kopt_forced_ocr"] = 0,
        ["doc_pages"] = 45,
        ["page_overlap_style"] = "dim",
        ["kopt_line_spacing"] = 1.2,
        ["normal_zoom_mode"] = "pagewidth",
        ["bbox"] = {},
        ["flipping_scroll_mode"] = false,
        ["percent_finished"] = 0.044444444444444,
        ["kopt_hw_dithering"] = 0,
        ["bookmarks_sorted"] = true
    },
    ["candidates"] = {
        [1] = {
            [1] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
            [2] = 1586887004
        },
        [2] = {
            [1] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua.old",
            [2] = 1586886864
        }
    },
    ["sidecar_file"] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
    ["history_file"] = "./history/[#mnt#Dane#koreader#test#] sample.pdf.lua",
    ["sidecar"] = "/mnt/Dane/koreader/test/sample.sdr",
    ["legacy_sidecar_file"] = "/mnt/Dane/koreader/test/sample.sdr/sample.pdf.lua"
}
```

**Comment by mwoz123** [comment]
@Frenzie - can you help with this status . where I can find it?
It's not in `docinfo.data`  (`DocSettings:open(self.ui.document.file).data`)
for both 100% read and unfinished

Logs: 

```
04/14/20-19:59:44 DEBUG Wallabag: self.ui.document.file = /mnt/Dane/koreader/test/sample.pdf
04/14/20-19:59:44 DEBUG data is read from  /mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua
04/14/20-19:59:44 DEBUG Wallabag: DocSettings:open(self.ui.document.file)  = {
    ["filepath"] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
    ["data"] = {
        ["kopt_word_spacing"] = -0.2,
        ["kopt_contrast"] = 1,
        ["kopt_text_wrap"] = 0,
        ["render_mode"] = 0,
        ["highlight_drawer"] = "lighten",
        ["kopt_screen_mode"] = "portrait",
        ["gamma"] = 1,
        ["kopt_page_margin"] = 0.1,
        ["screen_mode"] = "portrait",
        ["kopt_font_fine_tune"] = 0.05,
        ["inverse_reading_order"] = false,
        ["show_overlap_enable"] = false,
        ["kopt_defect_size"] = 1,
        ["stats"] = {
            ["performance_in_pages"] = {},
            ["md5"] = "dabfd5be70e60d0d464b64ded495a3be",
            ["authors"] = "",
            ["total_time_in_sec"] = 0,
            ["notes"] = 0,
            ["title"] = "sample",
            ["highlights"] = 0,
            ["pages"] = 45,
            ["series"] = "",
            ["language"] = ""
        },
        ["doc_props"] = {
            ["creationDate"] = "D:20140516224813+08'00'",
            ["keywords"] = "",
            ["producer"] = "itext-paulo-155 (itextpdf.sf.net-lowagie.com)",
            ["authors"] = "",
            ["author"] = "",
            ["subject"] = "",
            ["title"] = "sample",
            ["description"] = "",
            ["modDate"] = "D:20140516224813+08'00'",
            ["creator"] = "pdftk 2.01 - www.pdftk.com",
            ["series"] = "",
            ["language"] = ""
        },
        ["kopt_auto_straighten"] = 0,
        ["config_panel_index"] = 1,
        ["highlights_imported"] = true,
        ["kopt_page_scroll"] = 1,
        ["kopt_detect_indent"] = 1,
        ["page_positions"] = {
            [2] = 0.69696969696969,
            [1] = 0.85353535353535
        },
        ["kopt_quality"] = 1,
        ["kopt_writing_direction"] = 0,
        ["flipping_zoom_mode"] = "page",
        ["kopt_page_gap_height"] = 8,
        ["highlight"] = {},
        ["bookmarks"] = {},
        ["readermenu_tab_index"] = 7,
        ["kopt_doc_language"] = "eng",
        ["zoom_mode"] = "pagewidth",
        ["partial_md5_checksum"] = "f40c4da7f2861191c5ec19de4d80894f",
        ["kopt_page_opt"] = 0,
        ["disable_fuzzy_search"] = false,
        ["kopt_font_size"] = 1,
        ["kopt_full_screen"] = 1,
        ["kopt_trim_page"] = 1,
        ["last_page"] = 2,
        ["rotation_mode"] = 0,
        ["kopt_max_columns"] = 2,
        ["kopt_justification"] = 3,
        ["highlight_disabled"] = false,
        ["kopt_forced_ocr"] = 0,
        ["doc_pages"] = 45,
        ["page_overlap_style"] = "dim",
        ["kopt_line_spacing"] = 1.2,
        ["normal_zoom_mode"] = "pagewidth",
        ["bbox"] = {},
        ["flipping_scroll_mode"] = false,
        ["percent_finished"] = 0.044444444444444,
        ["kopt_hw_dithering"] = 0,
        ["bookmarks_sorted"] = true
    },
    ["candidates"] = {
        [1] = {
            [1] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
            [2] = 1586887004
        },
        [2] = {
            [1] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua.old",
            [2] = 1586886864
        }
    },
    ["sidecar_file"] = "/mnt/Dane/koreader/test/sample.sdr/metadata.pdf.lua",
    ["history_file"] = "./history/[#mnt#Dane#koreader#test#] sample.pdf.lua",
    ["sidecar"] = "/mnt/Dane/koreader/test/sample.sdr",
    ["legacy_sidecar_file"] = "/mnt/Dane/koreader/test/sample.sdr/sample.pdf.lua"
}
```

**Comment by Frenzie** [comment]
That's all already implemented in the plugin itself:

https://github.com/koreader/koreader/blob/25f224088ac5b819cab026c111b29940710d567d/plugins/wallabag.koplugin/main.lua#L667-L684

**Comment by mwoz123** [comment]
I've put my logic to https://github.com/koreader/koreader/pull/6048#L674

```
local percent_finished = docinfo.data.percent_finished
                    logger.dbg("Wallabag: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA status= ", status)
                    if status == "complete" or status == "abandoned" then
                        if self.remove_finished_from_history then
                            ReadHistory:removeItemByPath(entry_path)
                        end
                        if self.is_delete_finished then
                            self:removeArticle(entry_path)
                            num_deleted = num_deleted + 1
                        end
                    elseif percent_finished == 1 then -- 100% read
```

but it seems not to be executed :/

**Comment by mwoz123** [comment]
Cause doing this - this way 
```
function Wallabag:onCloseDocument()
    if self.remove_finished_from_history then
        local entry_path = self.ui.document.file
        if DocSettings:hasSidecarFile(entry_path) then
            local docinfo = DocSettings:open(entry_path)
            local status
            if docinfo.data.summary and docinfo.data.summary.status then
                status = docinfo.data.summary.status
            end
            logger.dbg("Wallabag: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA status= ", status)
            if status == "complete" or status == "abandoned" and self.remove_finished_from_history then
                ReadHistory:removeItemByPath(entry_path)
            end
        end
    end
end
```

also reports:
`04/14/20-20:44:42 DEBUG Wallabag: AAAAAAAAAAAAAAAAAAAAAAAAAAAAAA status= `

**Comment by Frenzie** [comment]
Line 675/`Wallabag:processLocalFiles(mode)` is for a sync with Wallabag, I merely mean that it does effectively the same thing, so it could be candidate for abstraction of common code or just for duplication.

Anyway, for this use case couldn't you just do something like self.ui.document:getLastPercent()?

**Comment by mwoz123** [comment]
requires: https://github.com/koreader/koreader-base/pull/1081

File: frontend/optmath.lua [hunk 6]
```diff
@@ -73,4 +73,8 @@ function Math.tmax(tab, func)
     return tmin_max(tab, func, "max")
 end
 
+function Math.roundPercent(percent)
```
Comment (Frenzie) [inline frontend/optmath.lua:76]: Please put it above by the other round functions instead of randomly at the bottom. ^_^

**Comment by poire-z** [comment]
Some luachecks warnings with the merged code (dunno if that will prevent next builds):
```
Luacheck results
Checking plugins/kosync.koplugin/main.lua         1 warning

    plugins/kosync.koplugin/main.lua:561:31: accessing undefined variable 'roundPercent'

Checking plugins/wallabag.koplugin/main.lua       1 warning

    plugins/wallabag.koplugin/main.lua:1046:36: line contains trailing whitespace

Total: 2 warnings / 0 errors in 425 files
```

**Pull request merged**